src/librustc_mir/build/scope.rs

```diff
@@ -591,21 +626,44 @@ fn build_scope_drops<'tcx>(cfg: &mut CFG<'tcx>,
                            -> BlockAnd<()> {
     let mut iter = scope.drops.iter().rev().peekable();
     while let Some(drop_data) = iter.next() {
-        // Try to find the next block with its cached block for us to diverge into in case the
-        // drop panics.
-        let on_diverge = iter.peek().iter().flat_map(|dd| dd.cached_block.into_iter()).next();
-        // If there’s no `cached_block`s within current scope, we must look for one in the
-        // enclosing scope.
-        let on_diverge = on_diverge.or_else(||{
-            earlier_scopes.iter().rev().flat_map(|s| s.cached_block()).next()
-        });
-        let next = cfg.start_new_block();
-        cfg.terminate(block, scope.source_info(drop_data.span), TerminatorKind::Drop {
-            location: drop_data.location.clone(),
-            target: next,
-            unwind: on_diverge
-        });
-        block = next;
+        let source_info = scope.source_info(drop_data.span);
+        if let DropKind::Value { .. } = drop_data.kind {
+            // Try to find the next block with its cached block
+            // for us to diverge into in case the drop panics.
+            let on_diverge = iter.peek().iter().filter_map(|dd| {
+                match dd.kind {
+                    DropKind::Value { cached_block } => cached_block,
+                    DropKind::Storage => None
+                }
+            }).next();
+            // If there’s no `cached_block`s within current scope,
+            // we must look for one in the enclosing scope.
+            let on_diverge = on_diverge.or_else(||{
+                earlier_scopes.iter().rev().flat_map(|s| s.cached_block()).next()
+            });
+            let next = cfg.start_new_block();
+            cfg.terminate(block, source_info, TerminatorKind::Drop {
+                location: drop_data.location.clone(),
+                target: next,
+                unwind: on_diverge
+            });
+            block = next;
+        }
+        match drop_data.kind {
+            DropKind::Value { .. } |
+            DropKind::Storage => {
```

Comment:
Why this `match` when this code should always execute? Just being defensive in case some variant is added in the future?